### PR TITLE
rgw_sal_motr: [CORTX-27628] Support Multipart Complete API in RGW

### PR DESCRIPTION
**Pull request by DPG17**
Handle deletion of old object during object overwrite situation, while executing Complete multipart API.

Signed-off-by: Dattaprasad Govekar <dattaprasad.govekar@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1561,6 +1551,16 @@ int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional
     return rc;
   }
 
+  // Finally, delete the object's entry from the bucket index.
```
Comment (DPG17) [inline src/rgw/rgw_sal_motr.cc:1554]: Only after deleting the old object (if any exists), delete entry from the bucket index

Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:1554]: For zero size object, we are not going to delete the motr object and we return at line number 1542. In that case we will not able to delete the metadata for zero size object. We need to remove the  `return` statement also make sure we don't try to delete the motr object for zero size object.

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -2950,6 +2951,39 @@ int MotrMultipartUpload::complete(const DoutPrefixProvider *dpp,
...
+
+  std::string obj_type = "simple object";
+  MotrObject::Meta old_meta;
+  rgw_bucket_dir_entry old_ent;
+
+  // Before updating bucket index with entry for new object, check if
+  // old object exists. Perform M0_IC_GET operation on bucket index.
+  rc = store->do_idx_op_by_name(bucket_index_iname, M0_IC_GET,
+                                target_obj->get_name(), old_check_bl);
+  if (rc == 0 && old_check_bl.length() > 0) {
```
Comment (sjain09) [inline src/rgw/rgw_sal_motr.cc:2963]: First check for bucket versioning status, if versioning is enabled then we should not delete old object.

**Review by sjain09 — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (DPG17) [reply src/rgw/rgw_sal_motr.cc:2963]: yes, correct

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (DPG17) [reply src/rgw/rgw_sal_motr.cc:1575]: done. Changes done to make sure S3 metadata for zero sized old object is deleted

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (DPG17) [reply src/rgw/rgw_sal_motr.cc:3019]: done

**Comment by DPG17** [comment]
Please note that Etag will still be missing in the Complete API response (after these code changes). To fix it, a code change has to happen in rgw_op.cc module. This needs to be handled by raising bug tracker on ceph's master branch in ceph community.

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -1550,7 +1550,29 @@ int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional
...
+  } else {
+    // Remove the motr object.
+    if (source->get_bucket()->get_info().versioning_status() == BUCKET_VERSIONED ||
+        source->get_bucket()->get_info().versioning_status() == BUCKET_VERSIONS_SUSPENDED) {
+          // TODO - Object Versioning feature
+          // If Bucket is versioned or versioning suspended, handle object deletion here.
+          // Presently, do nothing.
+          return 0;
+        }
+        else {
```
Comment (sjain09) [inline src/rgw/rgw_sal_motr.cc:1564]: Fix indentation for `else` block.

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -1550,7 +1550,29 @@ int MotrObject::MotrDeleteOp::delete_obj(const DoutPrefixProvider* dpp, optional
   // Delete from the cache first.
   source->store->get_obj_meta_cache()->remove(dpp, source->get_key().to_str());
 
-  // Delete the object's entry from the bucket index.
+  if (ent.meta.size == 0) {
+    ldpp_dout(dpp, 0) << __func__ << ": Object size is 0, not deleting motr object." << dendl;
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:1554]: make the debug level 20, 0 will keep logging this line everytime

**Review by bkirunge7 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 5]
```diff
@@ -3021,6 +3031,40 @@ int MotrMultipartUpload::complete(const DoutPrefixProvider *dpp,
...
+      old_mobj->set_category(RGWObjCategory::MultiMeta);
+    }
+    ldpp_dout(dpp, 20) << "MotrMultipartUpload::complete(): Old " << obj_type << " exists" << dendl;
+    // Delete old object
+    rc = old_mobj->delete_object(dpp, obj_ctx, y, true);
+    if (rc == 0) {
+      ldpp_dout(dpp, 20) << "MotrMultipartUpload::complete(): Old " << obj_type << " ["
+          << old_mobj->get_name() <<  "] deleted succesfully" << dendl;
+    } else {
+      ldpp_dout(dpp, 20) << "MotrMultipartUpload::complete(): Failed to delete old " << obj_type << " ["
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:3062]: As this is failure condition keep the log level to 0

Comment (DPG17) [reply src/rgw/rgw_sal_motr.cc]: done

**Review by sachinpunadikar — APPROVED** [review]
+1

File: src/rgw/rgw_sal_motr.cc [hunk 6]
```diff
@@ -3021,6 +3031,40 @@ int MotrMultipartUpload::complete(const DoutPrefixProvider *dpp,
...
+    ldpp_dout(dpp, 20) << "MotrMultipartUpload::complete(): Old " << obj_type << " exists" << dendl;
+    // Delete old object
+    rc = old_mobj->delete_object(dpp, obj_ctx, y, true);
+    if (rc == 0) {
+      ldpp_dout(dpp, 20) << "MotrMultipartUpload::complete(): Old " << obj_type << " ["
+          << old_mobj->get_name() <<  "] deleted succesfully" << dendl;
+    } else {
+      ldpp_dout(dpp, 0) << "MotrMultipartUpload::complete(): Failed to delete old " << obj_type << " ["
+          << old_mobj->get_name() <<  "]. Error = " << rc << dendl;
+      return rc;
```
Comment (sjain09) [inline src/rgw/rgw_sal_motr.cc:3064]: Should we return error to the user if we are not able to clean old object? Till this point we have uploaded the new object successfully. I think user should be impacted by the clean activity.

Comment (DPG17) [reply src/rgw/rgw_sal_motr.cc:3064]: no, we should not be returning error to the user in such case. This is similar to Cortx S3. Such condition will lead to object leak. However, it should be be part of Background delete handling - either using RGW GC or some other means(which is a separate topic/feature on "how to handle object leaks?"

Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:3064]: Yeah this will be handled with RGW GC, we can keep it for now.

**Review by sjain09 — APPROVED** [review]

**Pull request merged**